server/client_cloud.go

```diff
@@ -107,11 +107,12 @@ func (client jiraCloudClient) ListProjects(query string, limit int, expandIssueT
 			result.Values = result.Values[:remaining]
 		}
 		out = append(out, result.Values...)
-		remaining -= len(result.Values)
-
-		if !fetchAll && remaining == 0 {
-			// Got enough.
-			return out, nil
+		if !fetchAll {
+			remaining -= len(result.Values)
+			if remaining == 0 {
+				// Got enough.
+				return out, nil
+			}
 		}
```

Comment:
It took me a while to understand the function in general here, as I initially missed the re-assignment of `remaining` on line 84. It seems the bug we're experiencing only occurs when `fetchAll` is `true`. Is this correct?

At the code's current state, I'm thinking we should always pass in `50` to line 94. Otherwise, we will always fetch all that we want in one request, except for when we want all of the results. Also, a variable called `remaining` doesn't really make sense in the case of `fetchAll` being `true`, which makes this harder to read because of that as well. Idk if we really need to change anything there though.

Also, I'm not sure if line 106 will ever be `true`, since we always fetch a max of `remaining` at a given time. 

Since it seems your change fixes the underlying issue you're experiencing, I'm approving this, though wondering your opinion on changing line 94 to always provide `50` for `maxResults`.

@raghavaggarwal2308 @avas27JTG What are your thoughts on the above?

---

As an aside, just to make sure there is no bug in the implementation for Jira Server, I took a look at the `ListProjects` function there. There is no loop there, and we just ask the Jira server for whatever it will return, as `&jira.GetQueryOptions` does not allow for any `maxResults` or `limit` parameter, and then adjust the output based on the `limit` provided to our function. Another comment on this, we generally want to be more conscious of performance issues on Jira Server rather than Jira Cloud, since that more directly impacts the performance of the customer's infrastructure in general.

https://github.com/mattermost/mattermost-plugin-jira/blob/e1430fa88131f700ff91624fec61f325a912d419/server/client_server.go#L179-L197

Reply:
> Since it seems your change fixes the underlying issue you're experiencing, I'm approving this, though wondering your opinion on changing line 94 to always provide 50 for maxResults.

@mickmister I agree with your suggestion. I think we can send `50` in all the cases except when the value of `limit` is "<50".

> Also, I'm not sure if line 106 will ever be true, since we always fetch a max of remaining at a given time.

I also think this condition is unnecessary.

Reply:
Hi @raghavaggarwal2308, this solution was carried over from another user. While it's likely we can subtly improve the code, my team has been waiting a long time for a fix for this, so I would prefer we move forward at this point and defer any improvements to an additional iteration.

Reply:
@relwell Works for me 👍🏽.